pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java

```diff
@@ -130,6 +124,15 @@ public void start() {
     }
   }
 
+  private <T extends ServerBuilder<T>> Server buildGrpcServer(ServerBuilder<T> builder) {
+    return builder
+         // By using directExecutor, GRPC doesn't need to manage its own thread pool
```

Comment:
IIUC, this means that the request processing callbacks will happen directly on the transport thread - i.e., the Netty thread? And this is only recommended when the request processing is non-blocking - however, we still do proto deser on the request processing thread. I assume that that's okay because it should have minimal overhead and we don't really risk blocking Netty threads for handling other connections and requests?

Reply:
That deserialization is very cheap, non-blocking and it is expected to be done in the GRPC thread. There is no way to use direct mode without doing that deserialization.

Regarding Netty, that is a particularity of the transport. Still, this type of de/serialization is the kind of thing that is expected to be done in Netty threads in any Netty-based application.